torch/csrc/dynamo/guards.cpp

```diff
@@ -664,6 +664,615 @@ static struct PyModuleDef _module = {
+    } else {
+      throw py::type_error("LAMBDA_GUARD expects (callable, str)");
+    }
+  }
+
+  // Runs the lambda function with the current f_locals value.
+  bool check_nopybind(PyObject* value) override { // borrowed ref
+    PyObject* x = PyObject_CallOneArg(_guard_check_fn.ptr(), value); // new ref
+    bool result = PyObject_IsTrue(x);
+    Py_DECREF(x);
+    return result;
+  }
+
+ private:
+  // The user provided lambda function for check_fn.
+  py::function _guard_check_fn;
+};
+
+class TYPE_MATCH : public LeafGuard {
+ public:
+  // type_id = id(type(obj))
+  TYPE_MATCH(py::object type_id, py::object verbose_code_parts)
+      : LeafGuard(verbose_code_parts),
+        _expected(py::cast<unsigned long long>(type_id)) {}
+
+  bool check_nopybind(PyObject* value) override { // borrowed ref
+    return Py_TYPE(value) == (void*)_expected;
+  }
+
+ private:
+  // id of the type of the original object.
+  unsigned long long _expected;
+};
+
+class ID_MATCH : public LeafGuard {
+ public:
+  // obj_id = id(obj)
+  ID_MATCH(py::object obj_id, py::object verbose_code_parts)
+      : LeafGuard(verbose_code_parts),
+        _expected(py::cast<unsigned long long>(obj_id)) {}
+
+  bool check_nopybind(PyObject* value) override { // borrowed ref
+    return value == (void*)_expected;
+  }
+
+ private:
+  // id of the original object.
+  unsigned long long _expected;
+};
+
+/**
+ * Relational guards compare more than one value. We implement Relational
+ * guards by capturing some state in the guard object. For example for tensor
+ * aliasing guards - tensor X is not tensor Y - we construct one leaf guard
+ * and and install it at as a leaf of two guard managers (one for X and
+ * another for Y). Therefore, this guard is run twice. In the first
+ * invocation, it saves the first value (state) and returns True. In the
+ * second invocation, it compares the saved value with the new value and
+ * returns True if they do not alias.
+ *
+ * We have to be careful about resetting in case the other guards fail and we
+ * have some state in the relational guard. This is done by virtual method
+ * reset_state(). This is called by the GuardManager whenever
+ * there is a guard failure. In the event that the Guard evals to true, we do
+ * not need to reset the state. THe check_nopybind method should itself reset
+ * the state if it was called N times. So, fast path is unaffected.
+ *
+ * There is a question on which GuardManager node calls the
+ * reset_state. This is done by registering the guard as a
+ * relational_guard_resetter on the root node, which calls the resets all the
+ * relational guards on guard evaluation to False.
+ */
+class RelationalGuard : public LeafGuard {
+ public:
+  RelationalGuard(py::object verbose_code_parts)
+      : LeafGuard(verbose_code_parts) {}
+
+  // reset the relational guard state on guard failure. This is called by the
+  // guard manager.
+  virtual void reset_state() = 0;
```

Comment:
This won't be thread safe, so need to ensure we have a lock around guard checking.

Reply:
Wow! Thats a great catch.

Just so that I understand - This is not thread safe because two threads can try to run the same code object. So two threads might be running same guard function, and they can try to change the state simultaneously.

So, this is not just `reset_state`. This is bad even for `sorting` of guard managers where two threads might try to sort at the same time, and can leave the array in a bad state causing segfaults.

Added a std::lock_guard at the beginning of `check_nopybind` and `check_verbose_nopybind`. Please let me know if thats sufficient.

Reply:
The lock should be in the root guard manager.

Reply:
Yes, this is in RootGuardManager's `check_nopybind` and `check_verbose_nopybind`